### ISSUE TITLE
Problem using FormFyxer on a new machine still

### ISSUE THREAD

**Issue by nonprofittechy** [SuffolkLITLab/FormFyxer#33]
On this machine, I had already installed FormFyxer without getting it to work fully yet. Then I installed the latest version of FormFyxer from GitHub and tried to open the Weaver.

```
Error

[Retry](https://interviews.lemmalegal.com/interview?i=docassemble.ALWeaver%3Adata%2Fquestions%2Fassembly_line.yml&new_session=1#)

    OSError: [E050] Can't find model 'en_core_web_lg'. It doesn't seem to be a Python package or a valid path to a data directory.

Log

Traceback (most recent call last):
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/formfyxer/lit_explorer.py", line 36, in <module>
    nlp = spacy.load('en_core_web_lg') # this takes a while to loadimport os
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/spacy/__init__.py", line 51, in load
    return util.load_model(
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/spacy/util.py", line 427, in load_model
    raise IOError(Errors.E050.format(name=name))
OSError: [E050] Can't find model 'en_core_web_lg'. It doesn't seem to be a Python package or a valid path to a data directory.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/docassemble/webapp/server.py", line 7555, in index
    interview.assemble(user_dict, interview_status, old_user_dict, force_question=special_question)
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/docassemble/base/parse.py", line 7589, in assemble
    exec('import ' + module_name, user_dict)
  File "<string>", line 1, in <module>
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/formfyxer/__init__.py", line 1, in <module>
    from .lit_explorer import *
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/formfyxer/lit_explorer.py", line 42, in <module>
    nlp = spacy.load('en_core_web_lg') # this takes a while to loadimport os
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/spacy/__init__.py", line 51, in load
    return util.load_model(
  File "/usr/share/docassemble/local3.8/lib/python3.8/site-packages/spacy/util.py", line 427, in load_model
    raise IOError(Errors.E050.format(name=name))
OSError: [E050] Can't find model 'en_core_web_lg'. It doesn't seem to be a Python package or a valid path to a data directory.
```

**Comment by nonprofittechy** [comment]
I ended up manually logging in to the container to install the model from the commandline. Then I had to do a `supervisorctl restart uwsgi`

**Comment by nonprofittechy** [comment]
I think we should really find a way to make this fail more gracefully, too--many people might not need to use the ML capabilities of the Weaver.

**Comment by BryceStevenWilley** [comment]
Overall, I agree that we need a more graceful failure here. 

Specifically, I could correctly install from scratch locally at least once. I'll try a few more times, but I'm fairly confident I can reproduce a correct install on a localhost docker multiple times. But I wonder if lightsail servers with S3 need additional file sync, and how exactly that works. Just a guess about that though.

**Comment by nonprofittechy** [comment]
It's possible there was some complex interaction going on with some part of the system installed but not all of it? Not sure. All of the systems that I've tested on have been S3 though, and it worked on the apps-test server without an issue.

**Comment by BryceStevenWilley** [comment]
I did a little bit deeper of a dive into if it's possible to include the spacy model as a direct dependency, just to avoid all of these issues. The short answer is no, the longer answer is kinda: you can include it in the [requirements.txt](https://spacy.io/usage/models#models-download) like `https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.0.0/en_core_web_sm-3.0.0.tar.gz#egg=en_core_web_sm`, but docassemble (I think) only uses `setup.py`. To use `setup.py`, we can use PEP 508 which allows [URLs in the dependencies](https://stackoverflow.com/a/53383704/11416267), but pypi explicitly isn't going to support that for things not on Pypi. 

For many different reasons, [spacy won't be hosting the model packages on Pypi](https://github.com/explosion/spaCy/issues/3536). The current approach is our best bet, and seems to be what [others](https://github.com/allenai/allennlp/pull/933/files) have landed on as well.

I tried again, but couldn't reproduce on a local server. It took ~30 seconds to download the large model, maybe it takes much longer on AWS servers. Looking through the uwsgi.log would help there.

Will try to work on a good fallback for if it doesn't load next.

**Comment by nonprofittechy** [comment]
It's worth pursuing all of these leads. Some other thoughts if this turns out to be too much work:

1. Docassemble does include spacy already. I wonder if we can get an install method for models supported upstream, maybe through a similar system that installs debian packages on a stop/start?
2. Maybe we should ditch spacy? The packaging looks like it is always going to be hackish. @colarusso mentioned there are more modern alternatives but they might use more memory.
3. Maybe something we want to package ourselves in a separate docker container and call as a micro service? Doing a big download when the python module file loads is always going to be a bit tricky

**Comment by nonprofittechy** [comment]
Alternatively, we could have an install script that we require people to invoke manually as long as we can get it to work reliably when invoked from a Docassemble interview. We would just need to handle some way to warn the user that the module isn't installed.

**Comment by BryceStevenWilley** [comment]
This should be fixed by #44, but I'd like to make sure we don't hear of anyone having trouble installing the weaver for a few weeks before closing.
